### PR TITLE
Create a diagnostic AOV for invalid samples

### PR DESCRIPTION
**Pull request by oktomus**
This implements #1815.

### CODE REVIEW

File: src/appleseed/renderer/kernel/rendering/final/adaptivepixelrenderer.cpp [hunk 1]
```diff
@@ -73,6 +73,10 @@ namespace renderer
 
 namespace
 {
+
```
Comment (oktomus) [inline src/appleseed/renderer/kernel/rendering/final/adaptivepixelrenderer.cpp]: I will remove this line

File: src/appleseed/renderer/kernel/rendering/pixelrendererbase.h [hunk 2]
```diff
@@ -32,20 +32,26 @@
...
 #include <cstddef>
+#include <memory>
 
 // Forward declarations.
 namespace foundation    { class Tile; }
 namespace renderer      { class AOVAccumulatorContainer; }
 namespace renderer      { class Frame; }
 namespace renderer      { class TileStack; }
 
+using namespace foundation;
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/rendering/pixelrendererbase.h]: We completely forbid `using` declarations in header files since this defeats the point of namespaces entirely. Please use fully qualified names.

File: src/appleseed/renderer/kernel/rendering/pixelrendererbase.h [hunk 3]
```diff
@@ -32,20 +32,26 @@
 
 // appleseed.renderer headers.
 #include "renderer/kernel/rendering/ipixelrenderer.h"
+#include "renderer/modeling/frame/frame.h"
 
 // appleseed.foundation headers.
+#include "foundation/image/tile.h"
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/rendering/pixelrendererbase.h]: I presume you need this for `std::unique_ptr<>`? If you do, then please remove the `Tile` forward declaration below.

File: src/appleseed/renderer/kernel/rendering/pixelrendererbase.h [hunk 4]
```diff
@@ -32,20 +32,26 @@
 
 // appleseed.renderer headers.
 #include "renderer/kernel/rendering/ipixelrenderer.h"
+#include "renderer/modeling/frame/frame.h"
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/rendering/pixelrendererbase.h]: There's already a `Frame` forward declaration so this seems redundant.

File: src/appleseed/renderer/kernel/rendering/pixelrendererbase.h [hunk 5]
```diff
@@ -60,6 +66,8 @@ class PixelRendererBase
     // Constructor.
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/rendering/pixelrendererbase.h]: Change to plural now that there are two constructors.

File: src/appleseed/renderer/kernel/rendering/pixelrendererbase.h [hunk 6]
```diff
@@ -60,6 +66,8 @@ class PixelRendererBase
...
     PixelRendererBase();
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/rendering/pixelrendererbase.h]: Remove this blank line so that the comment applies to both constructors.

File: src/appleseed/renderer/kernel/rendering/pixelrendererbase.cpp [hunk 7]
```diff
@@ -51,22 +54,66 @@ PixelRendererBase::PixelRendererBase()
 {
 }
 
+PixelRendererBase::PixelRendererBase(
+    const Frame&        frame,
+    const size_t        thread_index)
+  : m_invalid_pixel_count(0)
+{
+    m_invalid_sample_aov_index = frame.create_extra_aov_image("invalid_samples");
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/rendering/pixelrendererbase.cpp]: The invalid samples AOV should be optional (since most of the time we don't need it, but it uses significant amounts of memory).

File: src/appleseed/renderer/kernel/rendering/pixelrendererbase.cpp [hunk 8]
```diff
@@ -51,22 +54,66 @@ PixelRendererBase::PixelRendererBase()
...
+
+    if ((thread_index == 0) && (m_invalid_sample_aov_index == size_t(~0)))
+    {
+        RENDERER_LOG_WARNING(
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/rendering/pixelrendererbase.cpp]: *invalid samples diagnostic* -> *invalid samples aov*

File: src/appleseed/renderer/kernel/rendering/pixelrendererbase.cpp [hunk 9]
```diff
@@ -51,22 +54,66 @@ PixelRendererBase::PixelRendererBase()
...
+                Color<uint8, 1> sample_state;
+                m_invalid_sample_diagnostic->get_pixel(x, y, sample_state);
+
+                Color4f color(1.0f, 0.0f, 1.0f, 1.0f);
+
+                if (sample_state[0] == CorrectSampleHint)
+                {
+                    tile.get_pixel(x, y, color);
+                    // 20% of luminance
+                    color.rgb().set(((color.r + color.g + color.b) / 3.0f) * 0.2f);
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/rendering/pixelrendererbase.cpp]: You can use `luminance(color)` which takes into account human perception of colors.

File: src/appleseed/renderer/kernel/rendering/pixelrendererbase.cpp [hunk 10]
```diff
@@ -41,6 +42,8 @@ using namespace foundation;
 
 namespace renderer
 {
+const uint8 InvalidSampleHint = 1;
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/rendering/pixelrendererbase.cpp]: Move after the "PixelRendererBase class implementation" block comment since this is technically part of the implementation of `PixelRendererBase`.

File: src/appleseed/renderer/kernel/rendering/final/uniformpixelrenderer.h [hunk 11]
```diff
@@ -32,6 +32,7 @@
 
 // appleseed.renderer headers.
 #include "renderer/kernel/rendering/ipixelrenderer.h"
+#include "renderer/modeling/frame/frame.h"
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/rendering/final/uniformpixelrenderer.h]: Prefer forward decls whenever possible (to reduce coupling and compile times).

File: src/appleseed/renderer/kernel/rendering/final/uniformpixelrenderer.cpp [hunk 12]
```diff
@@ -181,7 +183,7 @@ namespace
                     m_total_sampling_dim.insert(child_sampling_context.get_total_dimension());
 
                     // Merge the sample into the framebuffer.
-                    if (shading_result.is_valid())
+                    if (shading_result.is_valid() && shading_result.m_main.r < 0.8f)
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/rendering/final/uniformpixelrenderer.cpp]: Shouldn't be part of the PR :)

File: src/appleseed/renderer/kernel/rendering/final/adaptivepixelrenderer.cpp [hunk 13]
```diff
@@ -148,6 +150,7 @@ namespace
             Tile&                       tile,
             TileStack&                  aov_tiles) override
         {
+            PixelRendererBase::on_tile_end(frame, tile, aov_tiles);
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/rendering/final/adaptivepixelrenderer.cpp]: I'd suggest adding a blank line after this to make it stand out.

File: src/appleseed/renderer/kernel/rendering/final/adaptivepixelrenderer.cpp [hunk 14]
```diff
@@ -129,6 +130,7 @@ namespace
             Tile&                       tile,
             TileStack&                  aov_tiles) override
         {
+            PixelRendererBase::on_tile_begin(frame, tile, aov_tiles);
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/rendering/final/adaptivepixelrenderer.cpp]: I'd suggest adding a blank line after this to make it stand out.

File: src/appleseed/renderer/kernel/rendering/pixelrendererbase.h (hunk 2 above)
Comment (oktomus) [reply src/appleseed/renderer/kernel/rendering/pixelrendererbase.h]: That makes sense, we should change that also on the adaptive pixel renderer

File: src/appleseed/renderer/kernel/rendering/pixelrendererbase.h (hunk 4 above)
Comment (oktomus) [reply src/appleseed/renderer/kernel/rendering/pixelrendererbase.h]: Yes, my bad

File: src/appleseed/renderer/kernel/rendering/pixelrendererbase.cpp (hunk 7 above)
Comment (oktomus) [reply src/appleseed/renderer/kernel/rendering/pixelrendererbase.cpp]: @dictoon Is this possible to share some Parameters along pixel renderer base and sub classes ?

File: src/appleseed/renderer/kernel/rendering/final/uniformpixelrenderer.cpp (hunk 12 above)
Comment (oktomus) [reply src/appleseed/renderer/kernel/rendering/final/uniformpixelrenderer.cpp]: Woops, that's weird, I used `--patch` `to add and was careful.

**Comment by oktomus** [comment]
Thanks for the review @dictoon. Regarding different colors for different kind of errors, I was thinking of:
- Create another ShadingResult::is_valid so that we can get a error code (an enum telling what is the error)
- Don't extend the original one to avoid useless tests when we don't want the AOV

File: src/appleseed/renderer/kernel/rendering/pixelrendererbase.cpp (hunk 7 above)
Comment (oktomus) [reply src/appleseed/renderer/kernel/rendering/pixelrendererbase.cpp]: Or, I can use Frame::save_extra_aovs ? And we could do the same in the adaptive renderer ? 
Because `use diagnostics` and `save extra aovs` work only if they are both enabled.

Comment (dictoon) [reply src/appleseed/renderer/kernel/rendering/pixelrendererbase.cpp]: I think this coupling would be difficult to understand. In fact, do we agree that we have the same issue with the adaptive pixel sampler's diagnostic AOVs? Even if they are enabled, they aren't save to disk unless Save Extra AOVs is checked in the frame? If so then let's do the same for the invalid samples AOV, for now. (By the way, a checkbox must be added to the render settings to let users enable the invalid sample AOV.)

**Comment by dictoon** [comment]
> Create another ShadingResult::is_valid so that we can get a error code (an enum telling what is the error)

That sounds good, my only concerns are about complexity and especially performance. Let's try it (in a separate PR) and see how it goes!

File: src/appleseed/renderer/kernel/rendering/pixelrendererbase.cpp (hunk 7 above)
Comment (oktomus) [reply src/appleseed/renderer/kernel/rendering/pixelrendererbase.cpp]: Ok I'll do that !

File: src/appleseed/renderer/kernel/rendering/pixelrendererbase.h [hunk 15]
```diff
@@ -58,7 +62,7 @@ class PixelRendererBase
 {
   public:
     // Constructor.
-    PixelRendererBase();
+    PixelRendererBase(const Frame& frame, const size_t thread_index, const ParamArray& params);
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/rendering/pixelrendererbase.h]: You can make this more readable by having one parameter per line.

File: src/appleseed/renderer/kernel/rendering/pixelrendererbase.h [hunk 16]
```diff
@@ -74,18 +78,48 @@ class PixelRendererBase
...
+
+    size_t                                  m_invalid_sample_count;
+    size_t                                  m_invalid_pixel_count;
+    size_t                                  m_invalid_sample_aov_index;
+    std::unique_ptr<foundation::Tile>       m_invalid_sample_diagnostic;
+
+  protected:
+    const Parameters                m_base_params;
+};
+
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/rendering/pixelrendererbase.h]: Add a second blank line before block comments. That's the only case where we use two blank lines.

File: src/appleseed/renderer/kernel/rendering/pixelrendererbase.h [hunk 17]
```diff
@@ -74,18 +78,48 @@ class PixelRendererBase
...
+        AOVAccumulatorContainer&            aov_accumulators);
 
     void signal_invalid_sample();
 
   private:
-    size_t                          m_invalid_sample_count;
-    size_t                          m_invalid_pixel_count;
+    struct Parameters
+    {
+        const bool                  m_diagnostics;
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/rendering/pixelrendererbase.h]: It would look _nicer_ to forget about trying to align this to anything. There's no hard rule here, we're clearly in the domain of aesthetics :)

File: src/appleseed/renderer/kernel/rendering/final/adaptivepixelrenderer.cpp [hunk 18]
```diff
@@ -86,10 +83,11 @@ namespace
             ISampleRendererFactory*     factory,
             const ParamArray&           params,
             const size_t                thread_index)
-          : m_params(params)
+          : PixelRendererBase(frame, thread_index, params)
+          , m_params(params)
           , m_sample_renderer(factory->create(thread_index))
         {
-            if (m_params.m_diagnostics)
+            if (m_base_params.m_diagnostics)
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/rendering/final/adaptivepixelrenderer.cpp]: It's quite problematic to have parameters in two places, `m_base_params` and `m_params`. I would suggest simply keeping `m_base_params` in the base class, and renaming it to `m_params`.

File: src/appleseed/renderer/kernel/rendering/final/adaptivepixelrenderer.cpp [hunk 19]
```diff
@@ -126,11 +124,13 @@ namespace
 
         void on_tile_begin(
             const Frame&                frame,
-            Tile&                       tile,
+            foundation::Tile&           tile,
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/rendering/final/adaptivepixelrenderer.cpp]: You're in the .cpp file and there's a `using namespace foundation` statement so you can remove the qualification here (and everywhere in the file).

File: src/appleseed/renderer/kernel/rendering/final/adaptivepixelrenderer.cpp [hunk 20]
```diff
@@ -139,16 +139,19 @@ namespace
                     frame.aov_images().size(),
                     frame.get_filter()));
 
-            if (m_params.m_diagnostics)
-                m_diagnostics.reset(new Tile(tile.get_width(), tile.get_height(), 2, PixelFormatFloat));
+            if (m_base_params.m_diagnostics)
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/rendering/final/adaptivepixelrenderer.cpp]: Wrap the body of multiline `if` statements with braces:

    if (m_base_params.m_diagnostics)
    {
        ...
    }

File: src/appleseed/renderer/kernel/rendering/pixelrendererbase.cpp [hunk 21]
```diff
@@ -33,48 +33,103 @@
...
+                m_invalid_sample_diagnostic->get_pixel(x, y, sample_state);
+
+                foundation::Color4f color(1.0f, 0.0f, 1.0f, 1.0f);
+
+                if (sample_state[0] == CorrectSampleHint)
+                {
+                    tile.get_pixel(x, y, color);
+                    // 20% of luminance
+                    color.rgb().set(foundation::luminance(color.rgb()));
+                    color.rgb().set(color.r * 0.2f);
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/rendering/pixelrendererbase.cpp]: That looks a bit weird. Why not simply

    color.rgb().set(0.2f * luminance(color.rgb()));

File: src/appleseed/renderer/kernel/rendering/pixelrendererbase.cpp [hunk 22]
```diff
@@ -107,4 +176,25 @@ void PixelRendererBase::signal_invalid_sample()
     ++m_invalid_sample_count;
 }
 
+//
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/rendering/pixelrendererbase.cpp]: Add another blank line here too.

File: src/appleseed/renderer/kernel/rendering/pixelrendererbase.cpp [hunk 23]
```diff
@@ -99,6 +157,20 @@ void PixelRendererBase::on_pixel_end(
         {
             RENDERER_LOG_WARNING("more invalid samples found, omitting warning messages for brevity.");
         }
+
+    }
+
+    // Invalid samples diagnostic
+    if (tile_bbox.contains(pt) && m_params.m_diagnostics)
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/rendering/pixelrendererbase.cpp]: It would make sense to swap the conditions to avoid calling `AABB::contains()` if `m_params.m_diagnostics` is false.

File: src/appleseed/renderer/kernel/rendering/final/uniformpixelrenderer.h [hunk 24]
```diff
@@ -52,11 +52,12 @@ namespace renderer
 //
 
 class UniformPixelRendererFactory
-  : public IPixelRendererFactory
+  : public PixelRendererBaseFactory
 {
   public:
     // Constructor.
     UniformPixelRendererFactory(
+        const Frame&                frame,
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/rendering/final/uniformpixelrenderer.h]: Need to forward declare `Frame`.

File: src/appleseed/renderer/kernel/rendering/final/adaptivepixelrenderer.h [hunk 25]
```diff
@@ -41,8 +41,7 @@
 #include <cstddef>
 
 // Forward declarations.
-namespace foundation    { class Dictionary; }
-namespace renderer      { class Frame; }
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/rendering/final/adaptivepixelrenderer.h]: You need to restore the forward decl of `Frame`.

File: src/appleseed/renderer/kernel/rendering/final/adaptivepixelrenderer.cpp [hunk 26]
```diff
@@ -139,16 +142,19 @@ namespace
                     frame.aov_images().size(),
                     frame.get_filter()));
 
-            if (m_params.m_diagnostics)
-                m_diagnostics.reset(new Tile(tile.get_width(), tile.get_height(), 2, PixelFormatFloat));
+            if (PixelRendererBase::m_params.m_diagnostics)
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/rendering/final/adaptivepixelrenderer.cpp]: Add braces around multiline body.

**Comment by oktomus** [comment]
Thanks, I'm happy to know that my code is of quality :smile:
